### PR TITLE
Patch structured image logic

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
If we try to index non-image content into a field that is an image pointer, a 500 internal error will be raised and the whole add_documents batch will fail.

* **What is the new behavior (if this is a feature change)?**
A proper 400 error is returned for the specific document. Here is the example return:
```{'errors': True,
 'processingTimeMs': 2.4784463750000008,
 'index_name': 'test-image',
 'items': [{'status': 400,
   '_id': '1',
   'message': 'Could not find image found at `test`. \nReason: Input str of test is not a local file or a valid url. If you are using Marqo Cloud, please note that images can only be downloaded from a URL and local files are not supported. If you are running Marqo in a Docker container, you will need to use a Docker volume so that your container can access host files. For more information, please refer to: https://docs.marqo.ai/2.11/Guides/Advanced-Usage/images/',
   'error': 'Could not find image found at `test`. \nReason: Input str of test is not a local file or a valid url. If you are using Marqo Cloud, please note that images can only be downloaded from a URL and local files are not supported. If you are running Marqo in a Docker container, you will need to use a Docker volume so that your container can access host files. For more information, please refer to: https://docs.marqo.ai/2.11/Guides/Advanced-Usage/images/',
   'code': 'invalid_argument'}]}
```

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
running

* **Related Python client changes** (link commit/PR here)
no

* **Related documentation changes** (link commit/PR here)
no

* **Other information**:
no

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/core/models/marqo_add_documents_response.py [hunk 1]
```diff
@@ -31,14 +18,47 @@ def get_header_dict(self) -> Dict[str, str]:
         }
 
 
-class MarqoAddDocumentsResponse(MarqoBaseModel):
+class MarqoBaseDocumentsResponse(MarqoBaseModel):
+    """A base documents API response model."""
+    _batch_response_stats: BatchResponseStats = Field(exclude=True, default_factory=BatchResponseStats)
+
+    def dict(self, *args, **kwargs) -> Dict[str, Any]:
```
Comment (farshidz) [inline src/marqo/core/models/marqo_add_documents_response.py:25]: This needs a unit test

File: src/marqo/core/models/marqo_add_documents_response.py [hunk 2]
```diff
@@ -31,14 +18,47 @@ def get_header_dict(self) -> Dict[str, str]:
...
+        """Setting default exclude to exclude _batch_response_stats from the response.
+
+        Setting Field(exclude=True) does not work for
```
Comment (farshidz) [inline src/marqo/core/models/marqo_add_documents_response.py:28]: Is this a known pydantic issue?

File: src/marqo/s2_inference/clip_utils.py [hunk 3]
```diff
@@ -129,6 +129,20 @@ def load_image_from_path(image_path: str, image_download_headers: dict, timeout_
...
 
+def validate_url(url: str) -> bool:
+    """Validate a URL to ensure it is a valid URL. Returns True if the URL is valid or the encoded URL is valid.
+    Args:
+        url (str): URL to validate.
+    Returns:
+        bool: True if the URL is valid, False otherwise.
+    """
+    if isinstance(url, str):
+        return validators.url(url) or validators.url(encode_url(url))
```
Comment (farshidz) [inline src/marqo/s2_inference/clip_utils.py:140]: What are we trying to cover with encode_url? Is there a risk of regression or breaking change?

File: src/marqo/tensor_search/add_docs.py [hunk 4]
```diff
@@ -61,7 +64,7 @@ def threaded_download_and_preprocess_images(allocated_docs: List[dict], image_re
             for field in list(doc):
                 if field not in tensor_fields:
                     continue
-                if isinstance(doc[field], str) and clip_utils._is_image(doc[field]):
+                if force_download or (isinstance(doc[field], str) and clip_utils._is_image(doc[field])):
```
Comment (farshidz) [inline src/marqo/tensor_search/add_docs.py:67]: String check seems redundant as it's covered by `is_image`

File: src/marqo/tensor_search/tensor_search.py [hunk 5]
```diff
@@ -758,7 +759,8 @@ def _add_documents_structured(config: Config, add_docs_params: AddDocsParams, ma
                                      f"Is this a new tensor field?")
 
                 if len(chunks) == 0:  # Not using existing tensors or didn't find it
-                    if isinstance(field_content, (str, Image.Image)):
+                    is_image_pointer_field = marqo_field.type == FieldType.ImagePointer
+                    if isinstance(field_content, str):
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py:763]: Why were we checking for Image.Image before?

File: src/marqo/tensor_search/tensor_search.py [hunk 6]
```diff
@@ -769,7 +771,7 @@ def _add_documents_structured(config: Config, add_docs_params: AddDocsParams, ma
                         # 5. load correct media type into memory -> PIL (images), videos (), audio (torchaudio)
                         # 6. if chunking -> then add the extra chunker
 
-                        if isinstance(field_content, str) and not _is_image(field_content):
+                        if not is_image_pointer_field:
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py:774]: Previous check also required a string?

File: tests/core/document/test_documents_common.py [hunk 7]
```diff
@@ -39,4 +43,63 @@ def test_translate_vespa_document_response_logging(self):
...
+                    "status": 200,
+                    "message": "message",
+                    "error": "error"
+                }
+            ],
+            "processingTimeMs": 0.0
+        }
+        self.assertEqual(expected_dict, returned_dict)
+
+    def test_marqoGetDocumentsByIdsResponseDictFormat(self):
```
Comment (farshidz) [inline tests/core/document/test_documents_common.py:93]: Hard to understand what we're testing. A short description would help for each test

File: tests/s2_inference/test_clip_utils.py [hunk 8]
```diff
@@ -69,6 +69,8 @@ def test_is_image(self):
             ("Invalid PDF", 'document.pdf', False),
             ("Invalid TXT", 'text.txt', False),
             ("No extension", 'imagewithoutextension', False),
+            ("Valid URL with unencoded characters",
```
Comment (farshidz) [inline tests/s2_inference/test_clip_utils.py]: What's unencoded? Space?

File: tests/tensor_search/integ_tests/test_add_documents_combined.py [hunk 9]
```diff
@@ -65,7 +65,8 @@ def test_add_documents_with_truncated_image(self):
 
         documents = [
             {
-                "image_field_1": "https://marqo-assets.s3.amazonaws.com/tests/images/ai_hippo_statue.png",
+                "image_field_1": "https://raw.githubusercontent.com/marqo-ai/marqo/mainline"
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_add_documents_combined.py:68]: What's the reason for this change?

File: tests/tensor_search/integ_tests/test_add_documents_structured.py [hunk 10]
```diff
@@ -915,4 +915,26 @@ def test_long_double_numeric_values_edge_case(self):
...
                 )
-                self.assertEqual(expected_doc, returned_doc)
\ No newline at end of file
+                self.assertEqual(expected_doc, returned_doc)
+
+    def test_add_documents_nonImageContentForAnImageField(self):
+        """Test to ensure a proper error is raised when non-image content is added to an image field"""
+        documents = [
+            {
+                "_id": "1",
```
Comment (farshidz) [inline tests/tensor_search/integ_tests/test_add_documents_structured.py:928]: I think having 3 docs would make this more of an average case. I would probably also test having 1 doc with bad image URL, having 2 docs with bad image URLs, having multiple bad image URLs in one doc

**Review by farshidz — CHANGES_REQUESTED** [review]

File: src/marqo/core/models/marqo_add_documents_response.py (hunk 2 above)
Comment (wanliAlex) [reply src/marqo/core/models/marqo_add_documents_response.py:28]: We are using pydantic 1 so I can't find anything about this. I did a test on Pydantic 2 and the bug is resolved

File: src/marqo/core/models/marqo_add_documents_response.py (hunk 1 above)
Comment (wanliAlex) [reply src/marqo/core/models/marqo_add_documents_response.py:25]: Added

File: tests/core/document/test_documents_common.py (hunk 7 above)
Comment (wanliAlex) [reply tests/core/document/test_documents_common.py:93]: Done

File: src/marqo/s2_inference/clip_utils.py (hunk 3 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/clip_utils.py:140]: A customer URL that is not recognized as a valid URL. I think this is the expected behaviour because we are encoding the URL when downloading the image with a URL. So we should apply the same encoding when validating the URL.

File: src/marqo/tensor_search/add_docs.py (hunk 4 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/add_docs.py:67]: But _is_image also accepts other types, List, Tensor etc

File: src/marqo/tensor_search/tensor_search.py (hunk 5 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py:763]: I think we wanted to support indexing images as bytes but the feature was parked

File: src/marqo/tensor_search/tensor_search.py (hunk 6 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py:774]: I think the str check is not needed here anymore as we removed the Image.Image

File: tests/s2_inference/test_clip_utils.py (hunk 8 above)
Comment (wanliAlex) [reply tests/s2_inference/test_clip_utils.py]: Yes

File: tests/tensor_search/integ_tests/test_add_documents_combined.py (hunk 9 above)
Comment (wanliAlex) [reply tests/tensor_search/integ_tests/test_add_documents_combined.py:68]: It is reported that this test fails occasionally. We believe it's a transient network issue. My experience is that GitHub is more reliable than s3 in unit tests.

File: tests/tensor_search/integ_tests/test_add_documents_structured.py (hunk 10 above)
Comment (wanliAlex) [reply tests/tensor_search/integ_tests/test_add_documents_structured.py:928]: Done

**Review by farshidz — DISMISSED** [review]

File: src/marqo/tensor_search/tensor_search.py [hunk 11]
```diff
@@ -1020,7 +1022,7 @@ def _add_documents_structured(config: Config, add_docs_params: AddDocsParams, ma
 
         marqo_add_documents_response = config.document.translate_add_documents_response(
             index_responses, index_name=add_docs_params.index_name, unsuccessful_docs=unsuccessful_docs,
-            add_docs_processing_time=t1 - t0
+            add_docs_processing_time_ms=1000 * (t1 - t0)
```
Comment (papa99do) [inline src/marqo/tensor_search/tensor_search.py:1025]: Is this a breaking change? A safer way is to keep both and mark `add_docs_processing_time` as deprecated to be removed in v3

**Review by papa99do — DISMISSED** [review]

**Review by papa99do — APPROVED** [review]
